### PR TITLE
Homework 4

### PR DESCRIPTION
**Pull request by ghvik**

### CODE REVIEW

File: D04_ex00.py [hunk 1]
```diff
@@ -11,19 +11,38 @@
...
+            user_input = int(input(prompt))
+        except ValueError:
+            print("Sorry, that's not a number. Please enter it using digits.")
+        else:
+            if num == user_input:
+                print("Congrats, you guessed it right!")
+                break
+            if n == 1:
+                # Do not give the user another hint if they are out of tries
+                continue
```
Comment (danielsgriffin) [inline D04_ex00.py]: Why are you using continue here?

File: D04_ex00.py [hunk 2]
```diff
@@ -11,19 +11,38 @@
...
+            .format(n, "tries" if n > 1 else "try"))
+        try:
+            user_input = int(input(prompt))
+        except ValueError:
+            print("Sorry, that's not a number. Please enter it using digits.")
+        else:
+            if num == user_input:
+                print("Congrats, you guessed it right!")
+                break
+            if n == 1:
```
Comment (danielsgriffin) [inline D04_ex00.py]: Why if here?

File: HW04_ch07_ex02.py [hunk 3]
```diff
@@ -21,15 +21,32 @@
 
 ###############################################################################
 # Imports
-
+import pdb
```
Comment (danielsgriffin) [inline HW04_ch07_ex02.py]: ?

File: HW04_ch07_ex02.py [hunk 4]
```diff
@@ -21,15 +21,32 @@
...
+        
+        # Evaluate result
+        if user_input != "done":
+            try:
+                result = eval(user_input)
+            except (SyntaxError, NameError) as e:
+                print("Please check if there are any syntax errors or undefined variables in your code.", e)
+            print(result)
+        else:
+            return result
```
Comment (danielsgriffin) [inline HW04_ch07_ex02.py]: Good.

File: HW04_ch08_ex00.py [hunk 5]
```diff
@@ -18,17 +18,21 @@
 ###############################################################################
 # Imports
 
-
 # Body
-
+def count(word, letter):
+    count = 0
+    for l in word:
```
Comment (danielsgriffin) [inline HW04_ch08_ex00.py]: l?! Or was that I?

https://www.python.org/dev/peps/pep-0008/#names-to-avoid

File: HW04_ch08_ex05.py [hunk 6]
```diff
@@ -2,3 +2,52 @@
 # See Chapter 8: Strings Exercise 5 for the task.
 # Please do provide function calls that test/demonstrate your
```
Comment (danielsgriffin) [inline HW04_ch08_ex05.py]:
```
Book Examples:
 OK  'A'     by 3   => got: 'D'     expected: 'D'
 OK  'Z'     by 1   => got: 'A'     expected: 'A'
 OK  'cheer' by 7   => got: 'jolly' expected: 'jolly'
 OK  'melon' by -10 => got: 'cubed' expected: 'cubed'
 OK  'IBM'   by -1  => got: 'HAL'   expected: 'HAL'

Daniel's Tests:
 OK  'abc'   by 1   => got: 'bcd'   expected: 'bcd'
  X  'xyz'   by 1   => got: 'y`a'   expected: 'yza'
 OK  'ABC'   by 1   => got: 'BCD'   expected: 'BCD'
  X  'XYZ'   by 1   => got: 'Y@A'   expected: 'YZA'
 OK  'Az'    by 1   => got: 'Ba'    expected: 'Ba'
 OK  'Aa'    by 26  => got: 'Aa'    expected: 'Aa'
  X  'abc'   by -1  => got: '`ab'   expected: 'zab'
 OK  'xyz'   by -1  => got: 'wxy'   expected: 'wxy'
  X  'ABC'   by -1  => got: '@AB'   expected: 'ZAB'
 OK  'XYZ'   by -1  => got: 'WXY'   expected: 'WXY'
  X  'Az'    by -1  => got: '@y'    expected: 'Zy'
 OK  'Aa'    by -26 => got: 'Aa'    expected: 'Aa'
```

File: HW04_ch08_ex05.py [hunk 7]
```diff
@@ -2,3 +2,52 @@
...
+def rotate_word(word, rotate_by):
+    """Returns an encrypted word by rotating each letter by some given
+    amount.
+    
+    Parameters:
+    word -- the original word to encrypt
+    rotate_by -- an integer amount by which to rotate each letter
+    """
+    
+    a = 97
```
Comment (danielsgriffin) [inline HW04_ch08_ex05.py]: Why did you hardcode these like this?

File: HW04_ch08_ex05.py [hunk 8]
```diff
@@ -2,3 +2,52 @@
...
+#
+# Caesar cypher
+#
+########################################################################
+
+def rotate_word(word, rotate_by):
+    """Returns an encrypted word by rotating each letter by some given
+    amount.
+    
+    Parameters:
```
Comment (danielsgriffin) [inline HW04_ch08_ex05.py]: Nice.

File: HW04_ch08_ex05.py [hunk 9]
```diff
@@ -2,3 +2,52 @@
...
+            new_letter = chr(new_pos)
+        else:
+            new_pos = (ord(letter) + (rotate_by)) % upper_z
+            if new_pos < upper_a:
+                new_pos += upper_a - 1
+            new_letter = chr(new_pos)
+        new_word += new_letter
+    return new_word
+
+def main():
```
Comment (danielsgriffin) [inline HW04_ch08_ex05.py]: I like your range of tests. Think about how you are imagining edge cases?

File: HW04_string2.py [hunk 10]
```diff
@@ -7,7 +7,7 @@
 # http://code.google.com/edu/languages/google-python-class/
 
 # Additional basic string exercises
-
+import pdb
```
Comment (danielsgriffin) [inline HW04_string2.py]: ?

**Comment by danielsgriffin** [comment]
Good job.

File: HW04_ch08_ex05.py (hunk 9 above)
Comment (carinasauter) [inline HW04_ch08_ex05.py]: I don't understand the need to introduce variable z. The numbers are hardcoded into the system so I think it would improve readability if you just wrote the value behind z.

Comment (carinasauter) [inline HW04_ch08_ex05.py]: You have two places (actually 3) in the code where you do  rotate_by % z / upper_z. I think that could be simplified a lot by doing this before even deciding whether it is upper case or lower case. I think you should be able to just use the modulo once in your entire code.

File: HW04_string2.py [hunk 11]
```diff
@@ -17,9 +17,13 @@
...
-    # +++your code here+++
-    return
-
+    if len(s) >= 3:
+        if s[-3:] == 'ing':
+            return s + 'ly'
+        else:
+            return s + 'ing'
+    else:
+        return s
```
Comment (muditk) [inline HW04_string2.py]: elegant!

File: HW04_string2.py [hunk 12]
```diff
@@ -30,8 +34,33 @@ def verbing(s):
...
+            bad_appears = True
+            break
+        bad_index += 1
+        bad_last_index += 1
+
+    # Check if both words appear in the correct order.
+    if not_appears and bad_appears and bad_index > not_last_index:
+        return s[:not_index] + "good" + s[bad_last_index:]
+    else:
+        return s
```
Comment (muditk) [inline HW04_string2.py]: I'm sure it works, but maybe rather longwinded? Just a stylistic note.
